File: Java/ProjetCovariance/src/graphe/Graphe.java
package graphe;

import java.util.HashSet;

public class Graphe {


	public HashSet<Sommet> ensembleSommets = new HashSet<Sommet>();
	public HashSet<Arete> ensembleAretes = new HashSet<Arete>();
	//Iterator<Arete> itArete = ensembleAretes.iterator();

	String nom;


	public Graphe(String nom){
		this.nom = nom;
	}


	public void addSommet(Sommet sommet){
		ensembleSommets.add(sommet);
		System.out.println("Ajout Sommet ok");
	}

	public void addArete(Arete arete)
	{
		if(this.ensembleSommets.isEmpty())
		{
			System.err.println("Le graphe ne comporte pas de sommets");
		}
		if((this.ensembleSommets.contains(arete.getSommet1()) && this.ensembleSommets.contains(arete.getSommet2())))
		{
			System.out.println("L'arrete a les memes sommets que le graphe");
			this.ensembleAretes.add(arete);
			arete.setGraphe(this);
		}
		else
		{
			System.err.println("Le graphe n'a pas les meme sommets que l'arrete");
		}
	}

	public void removeSommet(Sommet sommet)
	{
		if(ensembleSommets.remove(sommet))
		{
			System.out.println("Sommet ("+sommet.getX()+","+sommet.getY() +") suppression dans removeSommet");
			for(Arete ar:ensembleAretes)
			{
				if(ar.getSommet1().equals(sommet) || ar.getSommet2().equals(sommet))
				{
					ensembleAretes.remove(ar);
					System.out.println("Une arrete a été supprimée");
				}
			}
		}
		else
		{
			System.err.println("Sommet ("+sommet.getX()+","+sommet.getY()+") ne fait pas partie du graphe "+this.getNom());
		}
	}

	public void removeArete(Arete arete){
		if(this.ensembleAretes.remove(arete)){
			System.out.println("Arete suprrimé "+this.getNom() );
		}
		else{
			System.err.println("Arete ne fait pas partie du graphe "+this.getNom() );
		}
	}


	public boolean invarianceGraphe(){

		for(Arete ar:ensembleAretes){
			if(this != ar.graphe){
				System.err.println("Graphe n'est pas égal à un des graphes des aretes ");
				return false;
			}

		}
		System.out.println("Toutles les arretes ont le bon graphe");
		return true;

	}


	//Getters and setters


	public HashSet<Sommet> getEnsembleSommets() {
		return ensembleSommets;
	}

	public void setEnsembleSommets(HashSet<Sommet> ensembleSommets) {
		this.ensembleSommets = ensembleSommets;
	}

	public HashSet<Arete> getEnsembleAretes() {
		return ensembleAretes;
	}

	public void setEnsembleAretes(HashSet<Arete> ensembleAretes) {
		this.ensembleAretes = ensembleAretes;
	}



	public String getNom() {
		return nom;
	}

	public void setNom(String nom) {
		this.nom = nom;
	}

	public void createArete(Sommet sommet1, Sommet sommet2)
	{
		if((sommet1 != null) && (sommet2 != null))
		{
			Arete arete = new Arete(sommet1,sommet2);
			this.addArete(arete);
			System.out.println("Arete a été créé avec succès");
		}
		else
		{
			System.err.println("ARete n'a pas pu être créé car un des sommets est null");
		}

	}






}

File: CPP/CPP-Covariance/TemplateCPP2/General/Graphe.cpp
#include "Graphe.h"
#include "Arete.h"
#include "Sommet.h"
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
class Graphe;
using namespace std;

namespace general{

template <typename G, typename S, typename A>
  Graphe::Graphe ( ) {
    ensembleSommets = new vector<Sommet*>();
    ensembleAretes = new vector<Arete*>();
  }
template <typename G, typename S, typename A>

  Graphe::Graphe(string nom)
  {
    nomgraphe = nom;
    ensembleSommets = new vector<Sommet*>();
    ensembleAretes = new vector<Arete*>();
  }

template <typename G, typename S, typename A>


  Graphe::~Graphe ( ) {
  }

  Sommet* sommet2;

template <typename G, typename S, typename A>

  void Graphe::addSommet(S* sommet){
    ensembleSommets->push_back(sommet);
    cout <<"Sommet a été ajouté avec succès" << endl;
    cout << ensembleSommets->size()<< endl;
  }

template <typename G, typename S, typename A>

  void Graphe::addArrete(A* arete){

    if(ensembleSommets->empty())
      {
	cout << "Le graphe ne comporte pas de sommets" << endl;
      }

    cout << ensembleSommets->size()<< endl;


    if((find(ensembleSommets->begin(), ensembleSommets->end(), arete->getSommet1())) != ensembleSommets->end() && (find(ensembleSommets->begin(), ensembleSommets->end(), arete->getSommet2())) != ensembleSommets->end())
      {
	cout<< "L'arrete a les memes sommets que le graphe" <<endl;
	ensembleAretes->push_back(arete);
	arete->setGraphe(this);

      }
    else
      {
	cout<<"Le graphe n'a pas les meme sommets que l'arrete" <<endl;
      }



  }

template <typename G, typename S, typename A>
  void Graphe::removeSommet(S* sommet){

    if((find(ensembleSommets->begin(), ensembleSommets->end(), sommet)) != ensembleSommets->end()){
      cout<<"Sommet trouvé et va être supprimé" <<endl;
      vector<Sommet*>::iterator it = remove(ensembleSommets->begin(), ensembleSommets->end(), sommet);
      ensembleSommets->erase(it, ensembleSommets->end());

      cout << ensembleSommets->size()<< endl;

      typedef vector<Arete*>::iterator IS;
      vector<Arete*>::iterator itsupp;
      for (IS i = ensembleAretes->begin(); i != ensembleAretes->end(); ++i) {
	cout << ensembleAretes->size()<< endl;


	Arete* arre = *i;
	if(arre->getSommet1() == sommet || arre->getSommet2() == sommet ){

	  itsupp = remove(ensembleAretes->begin(), ensembleAretes->end(), arre);


	}

      }
      ensembleAretes->erase(itsupp, ensembleAretes->end());
      cout<<"Une arete a été supprimée" <<endl;

      cout << ensembleAretes->size()<< endl;

    }
    else
      {
	cout<< "Sommet ne fait pas partie du graphe" <<endl;
      }


  }

template <typename G, typename S, typename A>

  void Graphe::removeArete(A* arete){

    cout << ensembleAretes->size()<< endl;
    if((find(ensembleAretes->begin(), ensembleAretes->end(), arete)) != ensembleAretes->end()){
      cout<<"Arete trouvé et va être supprimé" <<endl;
      vector<Arete*>::iterator it = remove(ensembleAretes->begin(), ensembleAretes->end(), arete);
      ensembleAretes->erase(it, ensembleAretes->end());

      cout << ensembleAretes->size()<< endl;

    }
    else{
      cout<< "Arete ne fait pas partie du graphe" <<endl;
    }

  }

  bool Graphe::invarianceGraphe(){


    typedef vector<Arete*>::iterator IS;
    vector<Arete*>::iterator itsupp;
    for (IS i = ensembleAretes->begin(); i != ensembleAretes->end(); ++i) {
      cout << ensembleAretes->size()<< endl;
      cout <<"test"<< endl;


      Arete* arre = *i;
      if(arre->getGraphe() != this){

	cout << "Graphe n'est pas égal à un des grapges des aretes" << endl;
	return false;

      }


    }
    cout << "Toutes les arretes ont le bon graphe" << endl;
    return true;


  }

template <typename G, typename S, typename A>
  void Graphe::createArete(S* sommet1, S* sommet2){

    if(sommet1 != NULL && sommet2 != NULL){

      Arete* arete = new Arete(sommet1, sommet2);
      this->addArrete(arete);
      cout <<"Arete ajouté" << endl;

    }
    else{
      cout <<"Arret n'a pas pu être créé, problème avec les sommets" << endl;
    }






  }

}



